File: sort/insertion_sort/go/insertion_sort.go
package insertionsort

// InsertionSort sorts an integer slice by using the insertion sort algorithm.
func InsertionSort(arr []int) {
	for i := 1; i < len(arr); i++ {
		val := arr[i]
		j := i
		for j > 0 && arr[j-1] > val {
			arr[j] = arr[j-1]
			j--
		}
		arr[j] = val
	}
}

File: sort/insertion_sort/scala/InsertionSort.scala
object InsertionSort {

  def insertionSort(arr: Array[Int]): Array[Int] = {
    for (i <- 1 to arr.length - 1) {
      val currentValue = arr(i)
      var j = i - 1
      while (j >= 0 && currentValue < arr(j)) {
        arr(j + 1) = arr(j)
        arr(j) = currentValue
        j = j - 1
      }
    }
    arr
  }

  def main(args: Array[String]): Unit = {
    val arr = Array(3, 1, 4, 5, 9, 2, 6, 8, 7, 10)
    val sortedArray = insertionSort(arr)
    println(sortedArray.mkString(", "))
  }
}
